
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Uncovering IS-mediated discord in bacterial genomes.">
      
      
      
        <link rel="canonical" href="https://tomdstanton.github.io/eris/reference/eris/scan/">
      
      
        <link rel="prev" href="../pan/">
      
      
        <link rel="next" href="../seq/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.16">
    
    
      
        <title>eris.scan - eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.7e37652d.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#eris.scan" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è" class="md-header__button md-logo" aria-label="eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              eris.scan
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="deep-purple" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/tomdstanton/eris" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    tomdstanton/eris
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è" class="md-nav__button md-logo" aria-label="eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/tomdstanton/eris" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    tomdstanton/eris
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1" checked>
        
          
          <label class="md-nav__link" for="__nav_2_1" id="__nav_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>eris
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2_1">
            <span class="md-nav__icon md-icon"></span>
            <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>eris
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    eris
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../alignment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>alignment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../alphabet/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>alphabet
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cli/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>cli
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../db/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>db
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../external/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>external
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../graph/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>graph
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>io
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../mapper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>mapper
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>pan
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>scan
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>scan
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#eris.scan" class="md-nav__link">
    <span class="md-ellipsis">
      scan
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.scan.Scanner" class="md-nav__link">
    <span class="md-ellipsis">
      Scanner
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Scanner">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eris.scan.Scanner.scan" class="md-nav__link">
    <span class="md-ellipsis">
      scan
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>seq
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>utils
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#eris.scan" class="md-nav__link">
    <span class="md-ellipsis">
      scan
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.scan.Scanner" class="md-nav__link">
    <span class="md-ellipsis">
      Scanner
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Scanner">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eris.scan.Scanner.scan" class="md-nav__link">
    <span class="md-ellipsis">
      scan
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



  <h1><code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>scan</h1>

<div class="doc doc-object doc-module">



<a id="eris.scan"></a>
    <div class="doc doc-contents first">

        <p>Module for running the scan pipeline on bacterial genomes.</p>









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="eris.scan.Scanner" class="doc doc-heading">
            <code>Scanner</code>


</h2>


    <div class="doc doc-contents ">


        <p>A class that runs the eris scan pipeline on bacterial genomes. The scan pipeline takes a single bacterial genome,
predicts ORFs if necessary, and uses Minimap2 to find IS elements with nucleotide-nucleotide alignment against
contigs. Then, ORFs belonging to- or flanked by- the IS elements are identified
(including potential graph traversal), and are returned in a result object for later reporting.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="eris.scan.Scanner.db">db</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="Database (eris.db.Database)" href="../db/#eris.db.Database">Database</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The ISFinder sequence eris.Database object.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.scan.Scanner.align_config">align_config</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="Minimap2AlignConfig


  
      dataclass
   (eris.external.Minimap2AlignConfig)" href="../external/#eris.external.Minimap2AlignConfig">Minimap2AlignConfig</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimap2 alignment configuration.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.scan.Scanner.index_config">index_config</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="Minimap2IndexConfig


  
      dataclass
   (eris.external.Minimap2IndexConfig)" href="../external/#eris.external.Minimap2IndexConfig">Minimap2IndexConfig</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimap2 index configuration.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.scan.Scanner.gene_finder_config">gene_finder_config</span></code></td>
            <td>
                  <code><span title="eris.io.GeneFinderConfig">GeneFinderConfig</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Configuration for the pyrodigal.GeneFinder instance if needed.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.scan.Scanner.min_overlap">min_overlap</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-external" href="https://docs.python.org/3/library/functions.html#float">float</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum %overlap for a CDS to be part of an element.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.scan.Scanner.min_proximity">min_proximity</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-external" href="https://docs.python.org/3/library/functions.html#int">int</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum distance in bp for an element to have an effect on a CDS.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">eris.scanner</span><span class="w"> </span><span class="kn">import</span> <span class="n">Scanner</span>
<span class="gp">... </span><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="gp">... </span><span class="k">with</span> <span class="n">Scanner</span><span class="p">()</span> <span class="k">as</span> <span class="n">scanner</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">results</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">scanner</span><span class="o">.</span><span class="n">scan</span><span class="p">(</span><span class="o">*</span><span class="n">Path</span><span class="o">.</span><span class="n">iterdir</span><span class="p">(</span><span class="s1">&#39;genomes/&#39;</span><span class="p">)))</span>
</code></pre></div>







              <details class="quote">
                <summary>Source code in <code>eris/scan.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Scanner</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class that runs the eris scan pipeline on bacterial genomes. The scan pipeline takes a single bacterial genome,</span>
<span class="sd">    predicts ORFs if necessary, and uses Minimap2 to find IS elements with nucleotide-nucleotide alignment against</span>
<span class="sd">    contigs. Then, ORFs belonging to- or flanked by- the IS elements are identified</span>
<span class="sd">    (including potential graph traversal), and are returned in a result object for later reporting.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        db (Database): The ISFinder sequence eris.Database object.</span>
<span class="sd">        align_config (Minimap2AlignConfig): Minimap2 alignment configuration.</span>
<span class="sd">        index_config (Minimap2IndexConfig): Minimap2 index configuration.</span>
<span class="sd">        gene_finder_config (GeneFinderConfig): Configuration for the pyrodigal.GeneFinder instance if needed.</span>
<span class="sd">        min_overlap (float): Minimum %overlap for a CDS to be part of an element.</span>
<span class="sd">        min_proximity (int): Minimum distance in bp for an element to have an effect on a CDS.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; from eris.scanner import Scanner</span>
<span class="sd">        ... from pathlib import Path</span>
<span class="sd">        ... with Scanner() as scanner:</span>
<span class="sd">        ...     results = list(scanner.scan(*Path.iterdir(&#39;genomes/&#39;)))</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">align_config</span><span class="p">:</span> <span class="n">Minimap2AlignConfig</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">index_config</span><span class="p">:</span> <span class="n">Minimap2IndexConfig</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                 <span class="n">gene_finder_config</span><span class="p">:</span> <span class="n">GeneFinderConfig</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">pool</span><span class="p">:</span> <span class="s1">&#39;Executor&#39;</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">min_overlap</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">,</span>
                 <span class="n">min_proximity</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">150</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">:</span> <span class="n">Database</span> <span class="o">=</span> <span class="n">Database</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">align_config</span><span class="p">:</span> <span class="n">Minimap2AlignConfig</span> <span class="o">=</span> <span class="n">align_config</span> <span class="ow">or</span> <span class="n">Minimap2AlignConfig</span><span class="p">(</span><span class="n">c</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">index_config</span><span class="p">:</span> <span class="n">Minimap2IndexConfig</span> <span class="o">=</span> <span class="n">index_config</span> <span class="ow">or</span> <span class="n">Minimap2IndexConfig</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_finder_config</span><span class="p">:</span> <span class="n">GeneFinderConfig</span> <span class="o">=</span> <span class="n">gene_finder_config</span> <span class="ow">or</span> <span class="n">GeneFinderConfig</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">min_overlap</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="n">min_overlap</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">min_proximity</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">min_proximity</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_pool</span><span class="p">:</span> <span class="s1">&#39;Executor&#39;</span> <span class="o">=</span> <span class="n">pool</span>  <span class="c1"># Only init the pool if we need it for the gene finder</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_gene_finder</span><span class="p">:</span> <span class="s1">&#39;pyrodigal.GeneFinder&#39;</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="nd">@property</span>
    <span class="nd">@require</span><span class="p">(</span><span class="s1">&#39;pyrodigal&#39;</span><span class="p">)</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">gene_finder</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_gene_finder</span><span class="p">:</span>
            <span class="kn">from</span><span class="w"> </span><span class="nn">pyrodigal</span><span class="w"> </span><span class="kn">import</span> <span class="n">GeneFinder</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_gene_finder</span> <span class="o">=</span> <span class="n">GeneFinder</span><span class="p">(</span><span class="o">**</span><span class="n">asdict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_finder_config</span><span class="p">))</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_gene_finder</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">cleanup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_gene_finder</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__enter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__exit__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exc_type</span><span class="p">,</span> <span class="n">exc_val</span><span class="p">,</span> <span class="n">exc_tb</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__del__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pipeline</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">scan</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">genomes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">IO</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">,</span> <span class="n">Genome</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">ScannerResult</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Runs the pipeline on input genomes</span>

<span class="sd">        Arguments:</span>
<span class="sd">            genomes: Input genomes to scan, can be strings or Paths to files, SeqFile instances or Genome instances</span>

<span class="sd">        Yields:</span>
<span class="sd">            ScannerResult instances for each genome processed or None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">yield from</span> <span class="nb">map</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_pipeline</span><span class="p">,</span> <span class="n">genomes</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_pipeline</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">IO</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">,</span> <span class="n">Genome</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">ScannerResult</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">Genome</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">ScannerError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Could not parse </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2"> as Genome&quot;</span><span class="p">)</span> <span class="kn">from</span><span class="w"> </span><span class="nn">e</span>

        <span class="n">contigs_with_is</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Record</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="n">Minimap2</span><span class="p">(</span><span class="n">targets</span><span class="o">=</span><span class="n">genome</span><span class="p">,</span> <span class="n">index_config</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">index_config</span><span class="p">)</span> <span class="k">as</span> <span class="n">aligner</span><span class="p">:</span>  <span class="c1"># Align IS to contigs</span>
            <span class="k">for</span> <span class="n">contig</span><span class="p">,</span> <span class="n">alignments</span> <span class="ow">in</span> <span class="n">grouper</span><span class="p">(</span><span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">config</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">align_config</span><span class="p">),</span> <span class="s1">&#39;target&#39;</span><span class="p">):</span>
                <span class="p">(</span><span class="n">contig</span> <span class="o">:=</span> <span class="n">genome</span><span class="p">[</span><span class="n">contig</span><span class="p">])</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">as_feature</span><span class="p">(</span><span class="s1">&#39;mobile_element&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cull_all</span><span class="p">(</span><span class="n">alignments</span><span class="p">)))</span>
                <span class="n">contigs_with_is</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">contig</span><span class="p">)</span>  <span class="c1"># Add contig name to list, also indicator that we have alignments</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">contigs_with_is</span><span class="p">:</span>  <span class="c1"># No alignments found, warn and exit early</span>
            <span class="k">return</span> <span class="n">warn</span><span class="p">(</span><span class="s1">&#39;No alignments found&#39;</span><span class="p">,</span> <span class="n">ScannerWarning</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">genome</span><span class="o">.</span><span class="n">is_annotated</span><span class="p">:</span>  <span class="c1"># Get ORFs if necessary, only do this if we have alignments to exit early</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_finder</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">ScannerError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Could not predict ORFs in </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2">; is Pyrodigal installed?&quot;</span><span class="p">)</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">find_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_finder</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pool</span><span class="p">)}</span>
        <span class="k">else</span><span class="p">:</span>  <span class="c1"># Get the existing annotations from the genome, TODO: Should we support other feature types?</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="n">gene</span> <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">contig</span><span class="o">.</span><span class="n">features</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">kind</span> <span class="o">==</span> <span class="s1">&#39;CDS&#39;</span><span class="p">}</span>

        <span class="n">graph</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">as_feature_graph</span><span class="p">()</span>  <span class="c1"># Get the gene graph for traversing connected genes</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">ScannerResult</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>  <span class="c1"># Init the result container</span>

        <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="n">contigs_with_is</span><span class="p">:</span>  <span class="c1"># Iterate over contigs with IS element alignments</span>
            <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">kind</span> <span class="o">==</span> <span class="s1">&#39;mobile_element&#39;</span><span class="p">,</span> <span class="n">contig</span><span class="p">):</span>  <span class="c1"># type: Feature</span>

                <span class="c1"># This block performs a graph traversal (BFS) starting from the neighbors of the IS element</span>
                <span class="c1"># to find all genes that are part of the IS element vs. those that are merely flanked by it.</span>
                <span class="n">element</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">contig</span><span class="p">,</span> <span class="n">store_seq</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># Extract and store nucleotide sequence of element</span>
                <span class="n">element</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">[</span><span class="n">element</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">qualifiers</span>  <span class="c1"># Pull over the qualifiers from the database</span>
                <span class="n">element</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;mobile_element_type&#39;</span><span class="p">,</span> <span class="s1">&#39;insertion sequence&#39;</span><span class="p">))</span>
                <span class="n">result</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">element</span><span class="p">)</span>  <span class="c1"># Add element to result features</span>
                <span class="n">result</span><span class="o">.</span><span class="n">contexts</span><span class="p">[</span><span class="n">element</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">Context</span><span class="p">(</span><span class="n">element</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

                <span class="n">promoter</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">for</span> <span class="n">promoter</span> <span class="ow">in</span> <span class="n">element</span><span class="o">.</span><span class="n">find_promoters</span><span class="p">(</span><span class="n">contig</span><span class="p">):</span>
                    <span class="n">result</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">promoter</span><span class="p">)</span>  <span class="c1"># Add element promoter to result features</span>
                    <span class="n">result</span><span class="o">.</span><span class="n">contexts</span><span class="p">[</span><span class="n">promoter</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">Context</span><span class="p">(</span><span class="n">element</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="s1">&#39;inside&#39;</span><span class="p">)</span>  <span class="c1"># Add promoter to result context</span>

                <span class="c1"># A queue for edges to visit. Start with the direct neighbors of the IS element.</span>
                <span class="c1"># We use a deque for efficient appends and pops from the left (FIFO) and  a set to keep track of</span>
                <span class="c1"># edges we&#39;ve already processed to avoid cycles and redundant work.</span>
                <span class="n">processed_edges</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">edges_to_visit</span> <span class="o">:=</span> <span class="n">deque</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">get_neighbors</span><span class="p">(</span><span class="n">element</span><span class="o">.</span><span class="n">id</span><span class="p">)))</span>

                <span class="k">while</span> <span class="n">edges_to_visit</span><span class="p">:</span>
                    <span class="n">edge</span> <span class="o">=</span> <span class="n">edges_to_visit</span><span class="o">.</span><span class="n">popleft</span><span class="p">()</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">gene</span> <span class="o">:=</span> <span class="n">genes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">to</span><span class="p">))</span> <span class="ow">or</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">contexts</span><span class="p">:</span>
                        <span class="k">continue</span>  <span class="c1"># Use result to check processed genes</span>

                    <span class="n">result</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>  <span class="c1"># Add edge to result for gfa output</span>
                    <span class="n">result</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>  <span class="c1"># Add gene to result features</span>
                    <span class="n">gene_context</span> <span class="o">=</span> <span class="n">Context</span><span class="p">(</span>
                        <span class="n">element</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="s1">&#39;flanking&#39;</span><span class="p">,</span>
                        <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;same strand&#39;</span> <span class="k">if</span> <span class="n">element</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span> <span class="k">else</span> <span class="s1">&#39;opposite strand&#39;</span>
                    <span class="p">)</span>
                    <span class="n">translation</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">parent</span><span class="o">=</span><span class="n">genome</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">parent_id</span><span class="p">])</span>  <span class="c1"># Extract translation</span>
                    <span class="c1"># Calculate the context of the gene relative to the element.</span>
                    <span class="c1"># A gene can only be part of the IS element if it&#39;s on the same contig.</span>
                    <span class="c1"># Otherwise, it&#39;s a flanking gene by definition.</span>
                    <span class="c1"># TODO: Add stitching logic for elements that span multiple contigs</span>
                    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">parent_id</span> <span class="o">==</span> <span class="n">element</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">parent_id</span><span class="p">:</span>  <span class="c1"># On the same contig</span>
                        <span class="k">if</span> <span class="n">element</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_overlap</span><span class="p">:</span>  <span class="c1"># This gene is part of the IS element</span>
                            <span class="n">gene_context</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="s1">&#39;inside&#39;</span>  <span class="c1"># Since it&#39;s part of the IS, we need to explore its</span>
                            <span class="c1"># neighbors as well to find the full extent of the element.</span>
                            <span class="k">for</span> <span class="n">next_edge</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">get_neighbors</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
                                <span class="k">if</span> <span class="n">next_edge</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">processed_edges</span><span class="p">:</span>
                                    <span class="n">processed_edges</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">next_edge</span><span class="p">)</span>
                                    <span class="n">edges_to_visit</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">next_edge</span><span class="p">)</span>

                        <span class="k">else</span><span class="p">:</span>  <span class="c1"># Non-overlapping gene on same contig</span>
                            <span class="k">if</span> <span class="n">element</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">:</span> <span class="c1"># Element comes after gene</span>
                                <span class="n">gene_context</span><span class="o">.</span><span class="n">distance</span> <span class="o">=</span> <span class="n">element</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span>
                                <span class="n">gene_context</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="s1">&#39;upstream&#39;</span> <span class="k">if</span> <span class="n">element</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="s1">&#39;downstream&#39;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">gene_context</span><span class="o">.</span><span class="n">distance</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">element</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span>
                                <span class="n">gene_context</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="s1">&#39;downstream&#39;</span> <span class="k">if</span> <span class="n">element</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="s1">&#39;upstream&#39;</span>

                            <span class="k">if</span> <span class="n">gene_context</span><span class="o">.</span><span class="n">distance</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_proximity</span><span class="p">:</span>  <span class="c1"># Predict effects</span>
                                <span class="k">if</span> <span class="n">promoter</span> <span class="ow">and</span> <span class="n">gene_context</span><span class="o">.</span><span class="n">location</span> <span class="o">==</span> <span class="s1">&#39;downstream&#39;</span> <span class="ow">and</span> <span class="n">gene_context</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;same strand&#39;</span><span class="p">:</span>
                                    <span class="n">gene_context</span><span class="o">.</span><span class="n">effect</span> <span class="o">=</span> <span class="s1">&#39;upregulated&#39;</span>
                                <span class="k">elif</span> <span class="ow">not</span> <span class="n">gene</span><span class="o">.</span><span class="n">partial</span><span class="p">()</span> <span class="ow">and</span> <span class="p">(</span><span class="n">translation</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;V&#39;</span><span class="p">}</span> <span class="ow">or</span> <span class="n">translation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;*&#39;</span><span class="p">):</span>
                                    <span class="n">gene_context</span><span class="o">.</span><span class="n">effect</span> <span class="o">=</span> <span class="s1">&#39;disrupted&#39;</span>

                    <span class="n">result</span><span class="o">.</span><span class="n">contexts</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_context</span>

        <span class="n">result</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="s1">&#39;location.parent_id&#39;</span><span class="p">,</span> <span class="s1">&#39;location.start&#39;</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">result</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="eris.scan.Scanner.scan" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">scan</span><span class="p">(</span><span class="o">*</span><span class="n">genomes</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Runs the pipeline on input genomes</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>genomes</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-external" title="typing.Union" href="https://docs.python.org/3/library/typing.html#typing.Union">Union</a>[<a class="autorefs autorefs-external" href="https://docs.python.org/3/library/stdtypes.html#str">str</a>, <a class="autorefs autorefs-external" title="pathlib.Path" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path">Path</a>, <a class="autorefs autorefs-external" title="typing.IO" href="https://docs.python.org/3/library/typing.html#typing.IO">IO</a>, <a class="autorefs autorefs-internal" title="SeqFile (eris.io.SeqFile)" href="../io/#eris.io.SeqFile">SeqFile</a>, <a class="autorefs autorefs-internal" title="Genome (eris.io.Genome)" href="../io/#eris.io.Genome">Genome</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input genomes to scan, can be strings or Paths to files, SeqFile instances or Genome instances</p>
              </div>
            </td>
            <td>
                  <code>()</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Yields:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="eris.scan.ScannerResult">ScannerResult</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ScannerResult instances for each genome processed or None</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>eris/scan.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">scan</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">genomes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">IO</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">,</span> <span class="n">Genome</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">ScannerResult</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Runs the pipeline on input genomes</span>

<span class="sd">    Arguments:</span>
<span class="sd">        genomes: Input genomes to scan, can be strings or Paths to files, SeqFile instances or Genome instances</span>

<span class="sd">    Yields:</span>
<span class="sd">        ScannerResult instances for each genome processed or None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">yield from</span> <span class="nb">map</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_pipeline</span><span class="p">,</span> <span class="n">genomes</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>







  
  






                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["toc.follow", "content.action.edit"], "search": "../../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.50899def.min.js"></script>
      
    
  </body>
</html>