
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Uncovering IS-mediated discord in bacterial genomes.">
      
      
      
        <link rel="canonical" href="https://tomdstanton.github.io/eris/reference/eris/io/">
      
      
        <link rel="prev" href="../graph/">
      
      
        <link rel="next" href="../mapper/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.16">
    
    
      
        <title>eris.io - eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.7e37652d.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#eris.io" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è" class="md-header__button md-logo" aria-label="eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              eris.io
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="deep-purple" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="deep-purple" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/tomdstanton/eris" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    tomdstanton/eris
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è" class="md-nav__button md-logo" aria-label="eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    eris üß¨üßû‚Äç‚ôÄüîÆÔ∏è
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/tomdstanton/eris" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    tomdstanton/eris
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1" checked>
        
          
          <label class="md-nav__link" for="__nav_2_1" id="__nav_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>eris
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2_1">
            <span class="md-nav__icon md-icon"></span>
            <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>eris
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    eris
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../alignment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>alignment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../alphabet/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>alphabet
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cli/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>cli
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../db/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>db
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../external/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>external
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../graph/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>graph
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>io
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>io
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#eris.io" class="md-nav__link">
    <span class="md-ellipsis">
      io
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.Genome" class="md-nav__link">
    <span class="md-ellipsis">
      Genome
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.as_assembly_graph" class="md-nav__link">
    <span class="md-ellipsis">
      as_assembly_graph
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.as_feature_graph" class="md-nav__link">
    <span class="md-ellipsis">
      as_feature_graph
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.find_genes" class="md-nav__link">
    <span class="md-ellipsis">
      find_genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.from_file" class="md-nav__link">
    <span class="md-ellipsis">
      from_file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.random" class="md-nav__link">
    <span class="md-ellipsis">
      random
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.ReadFile" class="md-nav__link">
    <span class="md-ellipsis">
      ReadFile
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ReadFile">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eris.io.ReadFile.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.ReadSet" class="md-nav__link">
    <span class="md-ellipsis">
      ReadSet
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.SeqFile" class="md-nav__link">
    <span class="md-ellipsis">
      SeqFile
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SeqFile">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      __iter__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.close" class="md-nav__link">
    <span class="md-ellipsis">
      close
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.peek" class="md-nav__link">
    <span class="md-ellipsis">
      peek
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.read" class="md-nav__link">
    <span class="md-ellipsis">
      read
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.rewind" class="md-nav__link">
    <span class="md-ellipsis">
      rewind
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.group_reads" class="md-nav__link">
    <span class="md-ellipsis">
      group_reads
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.parse" class="md-nav__link">
    <span class="md-ellipsis">
      parse
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../mapper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>mapper
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>pan
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../scan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>scan
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>seq
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>utils
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#eris.io" class="md-nav__link">
    <span class="md-ellipsis">
      io
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.Genome" class="md-nav__link">
    <span class="md-ellipsis">
      Genome
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.as_assembly_graph" class="md-nav__link">
    <span class="md-ellipsis">
      as_assembly_graph
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.as_feature_graph" class="md-nav__link">
    <span class="md-ellipsis">
      as_feature_graph
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.find_genes" class="md-nav__link">
    <span class="md-ellipsis">
      find_genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.from_file" class="md-nav__link">
    <span class="md-ellipsis">
      from_file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.Genome.random" class="md-nav__link">
    <span class="md-ellipsis">
      random
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.ReadFile" class="md-nav__link">
    <span class="md-ellipsis">
      ReadFile
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ReadFile">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eris.io.ReadFile.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.ReadSet" class="md-nav__link">
    <span class="md-ellipsis">
      ReadSet
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.SeqFile" class="md-nav__link">
    <span class="md-ellipsis">
      SeqFile
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SeqFile">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      __iter__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.close" class="md-nav__link">
    <span class="md-ellipsis">
      close
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.peek" class="md-nav__link">
    <span class="md-ellipsis">
      peek
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.read" class="md-nav__link">
    <span class="md-ellipsis">
      read
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#eris.io.SeqFile.rewind" class="md-nav__link">
    <span class="md-ellipsis">
      rewind
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.group_reads" class="md-nav__link">
    <span class="md-ellipsis">
      group_reads
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#eris.io.parse" class="md-nav__link">
    <span class="md-ellipsis">
      parse
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



  <h1><code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>io</h1>

<div class="doc doc-object doc-module">



<a id="eris.io"></a>
    <div class="doc doc-contents first">

        <p>Module for parsing and managing bacterial sequence files and data.</p>









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="eris.io.Genome" class="doc doc-heading">
            <code>Genome</code>


</h2>


    <div class="doc doc-contents ">


        <p>A class representing a single genome assembly in memory with contigs and potentially edges</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.Genome.id">id</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-external" href="https://docs.python.org/3/library/stdtypes.html#str">str</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The ID of the genome.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.Genome.contigs">contigs</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-external" href="https://docs.python.org/3/library/stdtypes.html#dict">dict</a>[<a class="autorefs autorefs-external" href="https://docs.python.org/3/library/stdtypes.html#str">str</a>:<a class="autorefs autorefs-internal" title="Record (eris.seq.Record)" href="../seq/#eris.seq.Record">Record</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary of contig IDs as keys and Record objects as values.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.Genome.edges">edges</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-external" href="https://docs.python.org/3/library/stdtypes.html#list">list</a>[<a class="autorefs autorefs-internal" title="Edge (eris.graph.Edge)" href="../graph/#eris.graph.Edge">Edge</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of Edges connecting the contigs.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.Genome.is_annotated">is_annotated</span></code></td>
            <td>
                  <code><a class="autorefs autorefs-external" href="https://docs.python.org/3/library/functions.html#bool">bool</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the genome has been loaded with annotations (e.g. from Genbank / GFF / BED).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>







              <details class="quote">
                <summary>Source code in <code>eris/io.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Genome</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class representing a single genome assembly in memory with contigs and potentially edges</span>

<span class="sd">    Attributes:</span>
<span class="sd">        id: The ID of the genome.</span>
<span class="sd">        contigs: A dictionary of contig IDs as keys and Record objects as values.</span>
<span class="sd">        edges: A list of Edges connecting the contigs.</span>
<span class="sd">        is_annotated: True if the genome has been loaded with annotations (e.g. from Genbank / GFF / BED).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">id_</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">contigs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">:</span> <span class="s1">&#39;Record&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">edges</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Edge</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Represents a single bacterial genome to be loaded into memory from a file&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">id_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">:</span> <span class="s1">&#39;Record&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">contigs</span> <span class="ow">or</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Edge</span><span class="p">]</span> <span class="o">=</span> <span class="n">edges</span> <span class="ow">or</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_annotated</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">item</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;Record&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="p">[</span><span class="n">item</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__format__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">__format_spec</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;fna&#39;</span><span class="p">,</span> <span class="s1">&#39;ffn&#39;</span><span class="p">,</span> <span class="s1">&#39;faa&#39;</span><span class="p">,</span> <span class="s1">&#39;bed&#39;</span><span class="p">,</span> <span class="s1">&#39;gfa&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">__format_spec</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="fm">__str__</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">__format_spec</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;fna&#39;</span><span class="p">,</span> <span class="s1">&#39;ffn&#39;</span><span class="p">,</span> <span class="s1">&#39;faa&#39;</span><span class="p">,</span> <span class="s1">&#39;bed&#39;</span><span class="p">}:</span>
            <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">__format_spec</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="k">elif</span> <span class="n">__format_spec</span> <span class="o">==</span> <span class="s1">&#39;gfa&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">__format_spec</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Invalid format: </span><span class="si">{</span><span class="n">__format_spec</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">from_file</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">IO</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">],</span> <span class="n">annotations</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads a genome from a file with optional annotations, e.g. a FASTA with a BED/GFF file.</span>
<span class="sd">        :param file: Path to file (str/Path) or a SeqFile instance.</span>
<span class="sd">        :param annotations: Optional annotations as a file path (str/Path) or SeqFile instance.</span>
<span class="sd">        :return: A Genome instance representing the genome in memory</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">file</span> <span class="o">=</span> <span class="n">SeqFile</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">)</span> <span class="k">else</span> <span class="n">file</span>
        <span class="bp">self</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">Record</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">file</span><span class="o">.</span><span class="n">format</span> <span class="o">==</span> <span class="s1">&#39;gfa&#39;</span> <span class="ow">and</span> <span class="nb">next</span><span class="p">((</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s1">&#39;topology&#39;</span><span class="p">),</span>
                                                     <span class="s1">&#39;linear&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;circular&#39;</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Edge</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>  <span class="c1"># Add self loop for circular genomes</span>
                    <span class="c1"># We assume a GFA file already has this edge but this may not be the case</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">Edge</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">annotations</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">format</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;gfa&#39;</span><span class="p">}:</span>
                <span class="k">raise</span> <span class="n">GenomeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Can only provide annotations to FASTA and GFA files, not </span><span class="si">{</span><span class="n">file</span><span class="o">.</span><span class="n">format</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">annotations</span> <span class="o">=</span> <span class="n">SeqFile</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">)</span> <span class="k">else</span> <span class="n">annotations</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">annotations</span><span class="o">.</span><span class="n">format</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;gff&#39;</span><span class="p">,</span> <span class="s1">&#39;bed&#39;</span><span class="p">}:</span>
                <span class="k">raise</span> <span class="n">GenomeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Annotations must be in GFF or BED format, not </span><span class="si">{</span><span class="n">annotations</span><span class="o">.</span><span class="n">format</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">contig_id</span><span class="p">,</span> <span class="n">features</span> <span class="ow">in</span> <span class="n">grouper</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="s1">&#39;location.parent_id&#39;</span><span class="p">):</span>  <span class="c1"># Sort by contig id</span>
                <span class="k">if</span> <span class="n">contig</span> <span class="o">:=</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">contig_id</span><span class="p">):</span>  <span class="c1"># type: Record</span>
                    <span class="n">contig</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span><span class="o">*</span><span class="n">features</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">is_annotated</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">random</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span> <span class="n">id_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Record</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">rng</span><span class="p">:</span> <span class="n">Random</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">n_contigs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">min_contigs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">max_contigs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">gc</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">min_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
            <span class="n">max_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000000</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates a random genome assembly for testing purposes.</span>

<span class="sd">        :param id_: ID of the genome, if not provided a random one will be generated</span>
<span class="sd">        :param genome: Initial genome record; if not provided a random one will be generated</span>
<span class="sd">        :param rng: Random number generator.</span>
<span class="sd">        :param n_contigs: Number of contigs in the assembly. If not provided, a random number of contigs will be generated.</span>
<span class="sd">        :param min_contigs: Minimum number of contigs if n_contigs is not specified.</span>
<span class="sd">        :param max_contigs: Maximum number of contigs if n_contigs is not specified.</span>
<span class="sd">        :param gc: GC content of the genome.</span>
<span class="sd">        :param length: Total length of the genome. If not provided, a random length will be generated.</span>
<span class="sd">        :param min_len: Minimum length of the genome if length is not specified.</span>
<span class="sd">        :param max_len: Maximum length of the genome if length is not specified.</span>
<span class="sd">        :return: A Genome instance representing the random assembly.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">rng</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">rng</span> <span class="o">=</span> <span class="n">RESOURCES</span><span class="o">.</span><span class="n">rng</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">id_</span> <span class="ow">or</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid4</span><span class="p">()),</span> <span class="p">{</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">(</span><span class="n">genome</span> <span class="ow">or</span> <span class="n">Record</span><span class="o">.</span><span class="n">random</span><span class="p">(</span>
            <span class="kc">None</span><span class="p">,</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">rng</span><span class="p">,</span> <span class="n">gc</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">min_len</span><span class="p">,</span> <span class="n">max_len</span><span class="p">))</span><span class="o">.</span><span class="n">shred</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span> <span class="n">n_contigs</span> <span class="ow">or</span> <span class="n">rng</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">min_contigs</span><span class="p">,</span> <span class="n">max_contigs</span><span class="p">))})</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">as_assembly_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directed</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Graph</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the assembly as a graph where contigs are nodes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">Graph</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="p">,</span> <span class="n">directed</span><span class="o">=</span><span class="n">directed</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">as_feature_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directed</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Graph</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the assembly as a graph where features are nodes.</span>
<span class="sd">        Adjacent features on the same contig are connected. Features at the termini</span>
<span class="sd">        of connected contigs are also connected.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">feature_graph</span> <span class="o">=</span> <span class="n">Graph</span><span class="p">(</span><span class="n">directed</span><span class="o">=</span><span class="n">directed</span><span class="p">)</span>

        <span class="c1"># 1. Add intra-contig edges (connecting adjacent features on the same contig)</span>
        <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">contig</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="c1"># This assumes contig.as_edge_list() correctly connects adjacent features</span>
                <span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">contig</span><span class="o">.</span><span class="n">as_edge_list</span><span class="p">():</span>
                    <span class="n">feature_graph</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>

        <span class="c1"># 2. Add inter-contig edges by iterating through all assembly connections</span>
        <span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">as_assembly_graph</span><span class="p">(</span><span class="n">directed</span><span class="o">=</span><span class="n">directed</span><span class="p">)</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
            <span class="c1"># Get the original &#39;from&#39; and &#39;to&#39; contigs for this edge</span>
            <span class="n">fr_contig_orig</span><span class="p">,</span> <span class="n">to_contig_orig</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">fr</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">to</span><span class="p">)</span>

            <span class="c1"># Skip if either contig is missing or has no features</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">fr_contig_orig</span> <span class="ow">and</span> <span class="n">fr_contig_orig</span><span class="o">.</span><span class="n">features</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="p">(</span><span class="n">to_contig_orig</span> <span class="ow">and</span> <span class="n">to_contig_orig</span><span class="o">.</span><span class="n">features</span><span class="p">):</span>
                <span class="k">continue</span>

            <span class="c1"># Determine the correct orientation based on the assembly edge attributes</span>
            <span class="n">fr_contig</span> <span class="o">=</span> <span class="n">fr_contig_orig</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span> <span class="k">if</span> <span class="n">edge</span><span class="o">.</span><span class="n">fr_attribute</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">fr_contig_orig</span>
            <span class="n">to_contig</span> <span class="o">=</span> <span class="n">to_contig_orig</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span> <span class="k">if</span> <span class="n">edge</span><span class="o">.</span><span class="n">to_attribute</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">to_contig_orig</span>

            <span class="c1"># Identify the terminal features on the correctly oriented contigs</span>
            <span class="n">fr_feature</span> <span class="o">=</span> <span class="n">fr_contig</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>  <span class="c1"># The last feature on the &#39;from&#39; contig</span>
            <span class="n">to_feature</span> <span class="o">=</span> <span class="n">to_contig</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># The first feature on the &#39;to&#39; contig</span>

            <span class="c1"># Calculate the gap distance between the features</span>
            <span class="n">distance</span> <span class="o">=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fr_contig</span><span class="p">)</span> <span class="o">-</span> <span class="n">fr_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="o">+</span> <span class="n">to_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span>

            <span class="n">feature_graph</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span>  <span class="c1"># Add the new edge connecting the two features</span>
                <span class="n">Edge</span><span class="p">(</span>
                    <span class="n">fr_feature</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="n">to_feature</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="n">fr_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                    <span class="n">to_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                    <span class="n">distance</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">feature_graph</span>

    <span class="nd">@require</span><span class="p">(</span><span class="s1">&#39;pyrodigal&#39;</span><span class="p">)</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">find_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_finder</span><span class="p">:</span> <span class="s1">&#39;pyrodigal.GeneFinder&#39;</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">pool</span><span class="p">:</span> <span class="s1">&#39;Executor&#39;</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                   <span class="n">config</span><span class="p">:</span> <span class="n">GeneFinderConfig</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">Feature</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Predicts ORFs in the Genome using Pyrodigal.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            gene_finder: An optional pre-trained pyrodigal.GeneFinder instance. If not provided, one will be</span>
<span class="sd">                         initialized using the config parameter.</span>
<span class="sd">            pool: An optional Executor (e.g., ThreadPoolExecutor or ProcessPoolExecutor)</span>
<span class="sd">                  to parallelize gene finding across contigs. If None, a ThreadPoolExecutor will be created.</span>
<span class="sd">            config: A GeneFinderConfig object to configure the Pyrodigal GeneFinder. Only used if gene_finder</span>
<span class="sd">                    is None.</span>

<span class="sd">        Yields:</span>
<span class="sd">            Feature objects representing the predicted CDS (Coding Sequence) genes.</span>

<span class="sd">        Notes:</span>
<span class="sd">            - Requires pyrodigal to be installed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">pyrodigal</span><span class="w"> </span><span class="kn">import</span> <span class="n">__version__</span> <span class="k">as</span> <span class="n">_pyrodigal_version</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">pyrodigal</span><span class="w"> </span><span class="kn">import</span> <span class="n">Gene</span>

        <span class="k">if</span> <span class="n">gene_finder</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="kn">from</span><span class="w"> </span><span class="nn">pyrodigal</span><span class="w"> </span><span class="kn">import</span> <span class="n">GeneFinder</span>
            <span class="n">gene_finder</span> <span class="o">=</span> <span class="n">GeneFinder</span><span class="p">(</span><span class="o">**</span><span class="n">asdict</span><span class="p">(</span><span class="n">config</span> <span class="ow">or</span> <span class="n">GeneFinderConfig</span><span class="p">()))</span>

        <span class="k">if</span> <span class="n">gene_finder</span><span class="o">.</span><span class="n">training_info</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">gene_finder</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">contig</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="k">if</span> <span class="n">pool</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">pool</span> <span class="o">=</span> <span class="n">RESOURCES</span><span class="o">.</span><span class="n">pool</span>

        <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># Gene counter</span>
        <span class="k">for</span> <span class="n">contig</span><span class="p">,</span> <span class="n">genes</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span>
                <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">contig</span><span class="p">:</span> <span class="n">gene_finder</span><span class="o">.</span><span class="n">find_genes</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">contig</span><span class="o">.</span><span class="n">seq</span><span class="p">)),</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="p">):</span>
            <span class="n">features</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>  <span class="c1"># type: Gene</span>
                <span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cds</span> <span class="o">:=</span> <span class="n">Feature</span><span class="p">(</span>
                    <span class="n">id_</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">contig</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">n</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">:</span><span class="s1">05d</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s1">&#39;CDS&#39;</span><span class="p">,</span> <span class="n">seq</span><span class="o">=</span><span class="n">Seq</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sequence</span><span class="p">(),</span> <span class="s1">&#39;DNA&#39;</span><span class="p">),</span>
                    <span class="n">location</span><span class="o">=</span><span class="n">Location</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">begin</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">partial_begin</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">partial_end</span><span class="p">,</span>
                                      <span class="n">contig</span><span class="o">.</span><span class="n">id</span><span class="p">),</span>
                    <span class="n">qualifiers</span><span class="o">=</span><span class="p">[</span>
                        <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;translation&#39;</span><span class="p">,</span> <span class="n">Seq</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">translate</span><span class="p">(),</span> <span class="s1">&#39;Amino&#39;</span><span class="p">)),</span>
                        <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;transl_table&#39;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">translation_table</span><span class="p">),</span>
                        <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;inference&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;ab initio prediction:Pyrodigal:</span><span class="si">{</span><span class="n">_pyrodigal_version</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">),</span>
                        <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;GC&#39;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">gc_cont</span><span class="p">),</span>
                        <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;rbs_motif&#39;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">rbs_motif</span><span class="p">),</span>
                        <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;rbs_spacer&#39;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">rbs_spacer</span><span class="p">)</span>
                    <span class="p">]</span>
                <span class="p">))</span>
                <span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># Increment the counter</span>
                <span class="k">yield</span> <span class="n">cds</span>  <span class="c1"># Yield the CDS for potential use</span>

            <span class="n">contig</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span><span class="o">*</span><span class="n">features</span><span class="p">)</span>  <span class="c1"># Use the add_features method to sort with existing features</span>

        <span class="k">if</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># If genes were found, mark the genome as annotated</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">is_annotated</span> <span class="o">=</span> <span class="kc">True</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="eris.io.Genome.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">id_</span><span class="p">,</span> <span class="n">contigs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">edges</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Represents a single bacterial genome to be loaded into memory from a file</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">id_</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">contigs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">:</span> <span class="s1">&#39;Record&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">edges</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Edge</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Represents a single bacterial genome to be loaded into memory from a file&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">id_</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">:</span> <span class="s1">&#39;Record&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">contigs</span> <span class="ow">or</span> <span class="p">{}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Edge</span><span class="p">]</span> <span class="o">=</span> <span class="n">edges</span> <span class="ow">or</span> <span class="p">[]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">is_annotated</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.Genome.as_assembly_graph" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">as_assembly_graph</span><span class="p">(</span><span class="n">directed</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Returns the assembly as a graph where contigs are nodes</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">as_assembly_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directed</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Graph</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the assembly as a graph where contigs are nodes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">Graph</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="p">,</span> <span class="n">directed</span><span class="o">=</span><span class="n">directed</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.Genome.as_feature_graph" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">as_feature_graph</span><span class="p">(</span><span class="n">directed</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Returns the assembly as a graph where features are nodes.
Adjacent features on the same contig are connected. Features at the termini
of connected contigs are also connected.</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">as_feature_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directed</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Graph</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the assembly as a graph where features are nodes.</span>
<span class="sd">    Adjacent features on the same contig are connected. Features at the termini</span>
<span class="sd">    of connected contigs are also connected.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">feature_graph</span> <span class="o">=</span> <span class="n">Graph</span><span class="p">(</span><span class="n">directed</span><span class="o">=</span><span class="n">directed</span><span class="p">)</span>

    <span class="c1"># 1. Add intra-contig edges (connecting adjacent features on the same contig)</span>
    <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">contig</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="c1"># This assumes contig.as_edge_list() correctly connects adjacent features</span>
            <span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">contig</span><span class="o">.</span><span class="n">as_edge_list</span><span class="p">():</span>
                <span class="n">feature_graph</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>

    <span class="c1"># 2. Add inter-contig edges by iterating through all assembly connections</span>
    <span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">as_assembly_graph</span><span class="p">(</span><span class="n">directed</span><span class="o">=</span><span class="n">directed</span><span class="p">)</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
        <span class="c1"># Get the original &#39;from&#39; and &#39;to&#39; contigs for this edge</span>
        <span class="n">fr_contig_orig</span><span class="p">,</span> <span class="n">to_contig_orig</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">fr</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">to</span><span class="p">)</span>

        <span class="c1"># Skip if either contig is missing or has no features</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">fr_contig_orig</span> <span class="ow">and</span> <span class="n">fr_contig_orig</span><span class="o">.</span><span class="n">features</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="p">(</span><span class="n">to_contig_orig</span> <span class="ow">and</span> <span class="n">to_contig_orig</span><span class="o">.</span><span class="n">features</span><span class="p">):</span>
            <span class="k">continue</span>

        <span class="c1"># Determine the correct orientation based on the assembly edge attributes</span>
        <span class="n">fr_contig</span> <span class="o">=</span> <span class="n">fr_contig_orig</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span> <span class="k">if</span> <span class="n">edge</span><span class="o">.</span><span class="n">fr_attribute</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">fr_contig_orig</span>
        <span class="n">to_contig</span> <span class="o">=</span> <span class="n">to_contig_orig</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span> <span class="k">if</span> <span class="n">edge</span><span class="o">.</span><span class="n">to_attribute</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">to_contig_orig</span>

        <span class="c1"># Identify the terminal features on the correctly oriented contigs</span>
        <span class="n">fr_feature</span> <span class="o">=</span> <span class="n">fr_contig</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>  <span class="c1"># The last feature on the &#39;from&#39; contig</span>
        <span class="n">to_feature</span> <span class="o">=</span> <span class="n">to_contig</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># The first feature on the &#39;to&#39; contig</span>

        <span class="c1"># Calculate the gap distance between the features</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fr_contig</span><span class="p">)</span> <span class="o">-</span> <span class="n">fr_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="o">+</span> <span class="n">to_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span>

        <span class="n">feature_graph</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span>  <span class="c1"># Add the new edge connecting the two features</span>
            <span class="n">Edge</span><span class="p">(</span>
                <span class="n">fr_feature</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="n">to_feature</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="n">fr_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                <span class="n">to_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                <span class="n">distance</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">feature_graph</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.Genome.find_genes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">find_genes</span><span class="p">(</span><span class="n">gene_finder</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pool</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">config</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Predicts ORFs in the Genome using Pyrodigal.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gene_finder</code>
            </td>
            <td>
                  <code><span title="pyrodigal.GeneFinder">GeneFinder</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An optional pre-trained pyrodigal.GeneFinder instance. If not provided, one will be
         initialized using the config parameter.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>pool</code>
            </td>
            <td>
                  <code><span title="Executor">Executor</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An optional Executor (e.g., ThreadPoolExecutor or ProcessPoolExecutor)
  to parallelize gene finding across contigs. If None, a ThreadPoolExecutor will be created.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>config</code>
            </td>
            <td>
                  <code><span title="eris.io.GeneFinderConfig">GeneFinderConfig</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A GeneFinderConfig object to configure the Pyrodigal GeneFinder. Only used if gene_finder
    is None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Yields:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="Feature (eris.seq.Feature)" href="../seq/#eris.seq.Feature">Feature</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Feature objects representing the predicted CDS (Coding Sequence) genes.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="notes" open>
  <summary>Notes</summary>
  <ul>
<li>Requires pyrodigal to be installed.</li>
</ul>
</details>

            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@require</span><span class="p">(</span><span class="s1">&#39;pyrodigal&#39;</span><span class="p">)</span>
<span class="k">def</span><span class="w"> </span><span class="nf">find_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_finder</span><span class="p">:</span> <span class="s1">&#39;pyrodigal.GeneFinder&#39;</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">pool</span><span class="p">:</span> <span class="s1">&#39;Executor&#39;</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
               <span class="n">config</span><span class="p">:</span> <span class="n">GeneFinderConfig</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">Feature</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Predicts ORFs in the Genome using Pyrodigal.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        gene_finder: An optional pre-trained pyrodigal.GeneFinder instance. If not provided, one will be</span>
<span class="sd">                     initialized using the config parameter.</span>
<span class="sd">        pool: An optional Executor (e.g., ThreadPoolExecutor or ProcessPoolExecutor)</span>
<span class="sd">              to parallelize gene finding across contigs. If None, a ThreadPoolExecutor will be created.</span>
<span class="sd">        config: A GeneFinderConfig object to configure the Pyrodigal GeneFinder. Only used if gene_finder</span>
<span class="sd">                is None.</span>

<span class="sd">    Yields:</span>
<span class="sd">        Feature objects representing the predicted CDS (Coding Sequence) genes.</span>

<span class="sd">    Notes:</span>
<span class="sd">        - Requires pyrodigal to be installed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">pyrodigal</span><span class="w"> </span><span class="kn">import</span> <span class="n">__version__</span> <span class="k">as</span> <span class="n">_pyrodigal_version</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">pyrodigal</span><span class="w"> </span><span class="kn">import</span> <span class="n">Gene</span>

    <span class="k">if</span> <span class="n">gene_finder</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">pyrodigal</span><span class="w"> </span><span class="kn">import</span> <span class="n">GeneFinder</span>
        <span class="n">gene_finder</span> <span class="o">=</span> <span class="n">GeneFinder</span><span class="p">(</span><span class="o">**</span><span class="n">asdict</span><span class="p">(</span><span class="n">config</span> <span class="ow">or</span> <span class="n">GeneFinderConfig</span><span class="p">()))</span>

    <span class="k">if</span> <span class="n">gene_finder</span><span class="o">.</span><span class="n">training_info</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">gene_finder</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">contig</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

    <span class="k">if</span> <span class="n">pool</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">pool</span> <span class="o">=</span> <span class="n">RESOURCES</span><span class="o">.</span><span class="n">pool</span>

    <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># Gene counter</span>
    <span class="k">for</span> <span class="n">contig</span><span class="p">,</span> <span class="n">genes</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span>
            <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">contig</span><span class="p">:</span> <span class="n">gene_finder</span><span class="o">.</span><span class="n">find_genes</span><span class="p">(</span><span class="nb">bytes</span><span class="p">(</span><span class="n">contig</span><span class="o">.</span><span class="n">seq</span><span class="p">)),</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="p">):</span>
        <span class="n">features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>  <span class="c1"># type: Gene</span>
            <span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cds</span> <span class="o">:=</span> <span class="n">Feature</span><span class="p">(</span>
                <span class="n">id_</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">contig</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">n</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">:</span><span class="s1">05d</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s1">&#39;CDS&#39;</span><span class="p">,</span> <span class="n">seq</span><span class="o">=</span><span class="n">Seq</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sequence</span><span class="p">(),</span> <span class="s1">&#39;DNA&#39;</span><span class="p">),</span>
                <span class="n">location</span><span class="o">=</span><span class="n">Location</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">begin</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">partial_begin</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">partial_end</span><span class="p">,</span>
                                  <span class="n">contig</span><span class="o">.</span><span class="n">id</span><span class="p">),</span>
                <span class="n">qualifiers</span><span class="o">=</span><span class="p">[</span>
                    <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;translation&#39;</span><span class="p">,</span> <span class="n">Seq</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">translate</span><span class="p">(),</span> <span class="s1">&#39;Amino&#39;</span><span class="p">)),</span>
                    <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;transl_table&#39;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">translation_table</span><span class="p">),</span>
                    <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;inference&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;ab initio prediction:Pyrodigal:</span><span class="si">{</span><span class="n">_pyrodigal_version</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">),</span>
                    <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;GC&#39;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">gc_cont</span><span class="p">),</span>
                    <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;rbs_motif&#39;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">rbs_motif</span><span class="p">),</span>
                    <span class="n">Qualifier</span><span class="p">(</span><span class="s1">&#39;rbs_spacer&#39;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">rbs_spacer</span><span class="p">)</span>
                <span class="p">]</span>
            <span class="p">))</span>
            <span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># Increment the counter</span>
            <span class="k">yield</span> <span class="n">cds</span>  <span class="c1"># Yield the CDS for potential use</span>

        <span class="n">contig</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span><span class="o">*</span><span class="n">features</span><span class="p">)</span>  <span class="c1"># Use the add_features method to sort with existing features</span>

    <span class="k">if</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># If genes were found, mark the genome as annotated</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_annotated</span> <span class="o">=</span> <span class="kc">True</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.Genome.from_file" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">from_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">annotations</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Loads a genome from a file with optional annotations, e.g. a FASTA with a BED/GFF file.
:param file: Path to file (str/Path) or a SeqFile instance.
:param annotations: Optional annotations as a file path (str/Path) or SeqFile instance.
:return: A Genome instance representing the genome in memory</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">from_file</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">IO</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">],</span> <span class="n">annotations</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads a genome from a file with optional annotations, e.g. a FASTA with a BED/GFF file.</span>
<span class="sd">    :param file: Path to file (str/Path) or a SeqFile instance.</span>
<span class="sd">    :param annotations: Optional annotations as a file path (str/Path) or SeqFile instance.</span>
<span class="sd">    :return: A Genome instance representing the genome in memory</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">file</span> <span class="o">=</span> <span class="n">SeqFile</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">)</span> <span class="k">else</span> <span class="n">file</span>
    <span class="bp">self</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">Record</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">file</span><span class="o">.</span><span class="n">format</span> <span class="o">==</span> <span class="s1">&#39;gfa&#39;</span> <span class="ow">and</span> <span class="nb">next</span><span class="p">((</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">qualifiers</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s1">&#39;topology&#39;</span><span class="p">),</span>
                                                 <span class="s1">&#39;linear&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;circular&#39;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Edge</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>  <span class="c1"># Add self loop for circular genomes</span>
                <span class="c1"># We assume a GFA file already has this edge but this may not be the case</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">Edge</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">annotations</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">format</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;gfa&#39;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="n">GenomeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Can only provide annotations to FASTA and GFA files, not </span><span class="si">{</span><span class="n">file</span><span class="o">.</span><span class="n">format</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">annotations</span> <span class="o">=</span> <span class="n">SeqFile</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">SeqFile</span><span class="p">)</span> <span class="k">else</span> <span class="n">annotations</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">annotations</span><span class="o">.</span><span class="n">format</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;gff&#39;</span><span class="p">,</span> <span class="s1">&#39;bed&#39;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="n">GenomeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Annotations must be in GFF or BED format, not </span><span class="si">{</span><span class="n">annotations</span><span class="o">.</span><span class="n">format</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">contig_id</span><span class="p">,</span> <span class="n">features</span> <span class="ow">in</span> <span class="n">grouper</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="s1">&#39;location.parent_id&#39;</span><span class="p">):</span>  <span class="c1"># Sort by contig id</span>
            <span class="k">if</span> <span class="n">contig</span> <span class="o">:=</span> <span class="bp">self</span><span class="o">.</span><span class="n">contigs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">contig_id</span><span class="p">):</span>  <span class="c1"># type: Record</span>
                <span class="n">contig</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span><span class="o">*</span><span class="n">features</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">is_annotated</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.Genome.random" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">random</span><span class="p">(</span><span class="n">id_</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n_contigs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_contigs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_contigs</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">gc</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_len</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">max_len</span><span class="o">=</span><span class="mi">5000000</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Generates a random genome assembly for testing purposes.</p>
<p>:param id_: ID of the genome, if not provided a random one will be generated
:param genome: Initial genome record; if not provided a random one will be generated
:param rng: Random number generator.
:param n_contigs: Number of contigs in the assembly. If not provided, a random number of contigs will be generated.
:param min_contigs: Minimum number of contigs if n_contigs is not specified.
:param max_contigs: Maximum number of contigs if n_contigs is not specified.
:param gc: GC content of the genome.
:param length: Total length of the genome. If not provided, a random length will be generated.
:param min_len: Minimum length of the genome if length is not specified.
:param max_len: Maximum length of the genome if length is not specified.
:return: A Genome instance representing the random assembly.</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">random</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span> <span class="n">id_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Record</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">rng</span><span class="p">:</span> <span class="n">Random</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">n_contigs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">min_contigs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">max_contigs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">gc</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">min_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
        <span class="n">max_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000000</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a random genome assembly for testing purposes.</span>

<span class="sd">    :param id_: ID of the genome, if not provided a random one will be generated</span>
<span class="sd">    :param genome: Initial genome record; if not provided a random one will be generated</span>
<span class="sd">    :param rng: Random number generator.</span>
<span class="sd">    :param n_contigs: Number of contigs in the assembly. If not provided, a random number of contigs will be generated.</span>
<span class="sd">    :param min_contigs: Minimum number of contigs if n_contigs is not specified.</span>
<span class="sd">    :param max_contigs: Maximum number of contigs if n_contigs is not specified.</span>
<span class="sd">    :param gc: GC content of the genome.</span>
<span class="sd">    :param length: Total length of the genome. If not provided, a random length will be generated.</span>
<span class="sd">    :param min_len: Minimum length of the genome if length is not specified.</span>
<span class="sd">    :param max_len: Maximum length of the genome if length is not specified.</span>
<span class="sd">    :return: A Genome instance representing the random assembly.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">rng</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">rng</span> <span class="o">=</span> <span class="n">RESOURCES</span><span class="o">.</span><span class="n">rng</span>
    <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">id_</span> <span class="ow">or</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid4</span><span class="p">()),</span> <span class="p">{</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">(</span><span class="n">genome</span> <span class="ow">or</span> <span class="n">Record</span><span class="o">.</span><span class="n">random</span><span class="p">(</span>
        <span class="kc">None</span><span class="p">,</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">rng</span><span class="p">,</span> <span class="n">gc</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">min_len</span><span class="p">,</span> <span class="n">max_len</span><span class="p">))</span><span class="o">.</span><span class="n">shred</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span> <span class="n">n_contigs</span> <span class="ow">or</span> <span class="n">rng</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">min_contigs</span><span class="p">,</span> <span class="n">max_contigs</span><span class="p">))})</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="eris.io.ReadFile" class="doc doc-heading">
            <code>ReadFile</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="SeqFile (eris.io.SeqFile)" href="#eris.io.SeqFile">SeqFile</a></code></p>


        <p>A subclass of :class:<code>~eris.core.io.SeqFile</code> that specifically handles a single file of sequencing reads
in FASTQ format.
This class extracts and stores read-specific information from the filename, and is designed to be used with the
:class:<code>~eris.core.io.ReadSet</code> class.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.ReadFile.sample_name">sample_name</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the sample from which the reads were derived.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.ReadFile.read_type">read_type</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The type of read ('short', 'illumina', 'long', 'pb', 'ont', or 'unknown').</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.ReadFile.sample_number">sample_number</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The sample number (for Illumina reads).</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.ReadFile.lane_number">lane_number</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The lane number (for Illumina reads).</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.ReadFile.read_number">read_number</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The read number (1 or 2 for paired-end reads).</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="eris.io.ReadFile.index_number">index_number</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The index number (for Illumina reads).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>







              <details class="quote">
                <summary>Source code in <code>eris/io.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">ReadFile</span><span class="p">(</span><span class="n">SeqFile</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A subclass of :class:`~eris.core.io.SeqFile` that specifically handles a single file of sequencing reads</span>
<span class="sd">    in FASTQ format.</span>
<span class="sd">    This class extracts and stores read-specific information from the filename, and is designed to be used with the</span>
<span class="sd">    :class:`~eris.core.io.ReadSet` class.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        sample_name: The name of the sample from which the reads were derived.</span>
<span class="sd">        read_type: The type of read (&#39;short&#39;, &#39;illumina&#39;, &#39;long&#39;, &#39;pb&#39;, &#39;ont&#39;, or &#39;unknown&#39;).</span>
<span class="sd">        sample_number: The sample number (for Illumina reads).</span>
<span class="sd">        lane_number: The lane number (for Illumina reads).</span>
<span class="sd">        read_number: The read number (1 or 2 for paired-end reads).</span>
<span class="sd">        index_number: The index number (for Illumina reads).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param file: Path to file (str/Path) or an IO stream (binary or text).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_read_type</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;unknown&#39;</span><span class="p">,</span> <span class="s1">&#39;short&#39;</span><span class="p">,</span> <span class="s1">&#39;illumina&#39;</span><span class="p">,</span> <span class="s1">&#39;long&#39;</span><span class="p">,</span> <span class="s1">&#39;pb&#39;</span><span class="p">,</span> <span class="s1">&#39;ont&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;unknown&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sample_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_lane_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_read_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_index_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">m</span> <span class="o">:=</span> <span class="n">_ILLUMINA_READ_REGEX</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;sample_name&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_read_type</span> <span class="o">=</span> <span class="s2">&quot;illumina&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_sample_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;sample_number&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_lane_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;lane_number&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_read_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;read_number&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_index_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;index_number&#39;</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">m</span> <span class="o">:=</span> <span class="n">_SHORT_READ_REGEX</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;sample_name&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_read_type</span> <span class="o">=</span> <span class="s2">&quot;short&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_read_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;read_number&#39;</span><span class="p">))</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">read_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_type</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">sample_number</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_number</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">lane_number</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_lane_number</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">read_number</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_number</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">index_number</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_index_number</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="eris.io.ReadFile.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>:param file: Path to file (str/Path) or an IO stream (binary or text).</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    :param file: Path to file (str/Path) or an IO stream (binary or text).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_read_type</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;unknown&#39;</span><span class="p">,</span> <span class="s1">&#39;short&#39;</span><span class="p">,</span> <span class="s1">&#39;illumina&#39;</span><span class="p">,</span> <span class="s1">&#39;long&#39;</span><span class="p">,</span> <span class="s1">&#39;pb&#39;</span><span class="p">,</span> <span class="s1">&#39;ont&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;unknown&#39;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_sample_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_lane_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_read_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_index_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">m</span> <span class="o">:=</span> <span class="n">_ILLUMINA_READ_REGEX</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;sample_name&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_read_type</span> <span class="o">=</span> <span class="s2">&quot;illumina&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sample_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;sample_number&#39;</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_lane_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;lane_number&#39;</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_read_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;read_number&#39;</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_index_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;index_number&#39;</span><span class="p">))</span>
    <span class="k">elif</span> <span class="n">m</span> <span class="o">:=</span> <span class="n">_SHORT_READ_REGEX</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;sample_name&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_read_type</span> <span class="o">=</span> <span class="s2">&quot;short&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_read_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;read_number&#39;</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="eris.io.ReadSet" class="doc doc-heading">
            <code>ReadSet</code>


</h2>


    <div class="doc doc-contents ">


        <p>Whilst still representing genomes, these are different from Genome instances as they will not hold
sequence information in memory and may consist of multiple files.</p>







              <details class="quote">
                <summary>Source code in <code>eris/io.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">ReadSet</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Whilst still representing genomes, these are different from Genome instances as they will not hold</span>
<span class="sd">    sequence information in memory and may consist of multiple files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">files</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">ReadFile</span><span class="p">],</span> <span class="n">id_</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">id_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_files</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">read_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">ReadFile</span><span class="p">):</span>
                <span class="n">file</span> <span class="o">=</span> <span class="n">ReadFile</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">sample_name</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">!=</span> <span class="n">file</span><span class="o">.</span><span class="n">sample_name</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">ReadSetError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">file</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1"> sample name </span><span class="si">{</span><span class="n">file</span><span class="o">.</span><span class="n">sample_name</span><span class="si">}</span><span class="s1"> does not match ReadSet </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="n">read_types</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">read_type</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_types</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ReadSetError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Hybrid read set </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1"> not yet supported&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_set_type</span> <span class="o">=</span> <span class="n">read_types</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_set_type</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_files</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">read</span><span class="p">)</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_files</span><span class="p">])</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_files</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="eris.io.SeqFile" class="doc doc-heading">
            <code>SeqFile</code>


</h2>


    <div class="doc doc-contents ">


        <p>Class for handling a (possibly compressed) file or stream of biological formats.</p>
<p>:param file: Path to file (str/Path) or an IO stream (binary or text).
:param format_: Format of the file. If None, the format will be guessed from the
                filename extension or the stream content.
:return: A SeqFile instance</p>







              <details class="quote">
                <summary>Source code in <code>eris/io.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">SeqFile</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class for handling a (possibly compressed) file or stream of biological formats.</span>

<span class="sd">    :param file: Path to file (str/Path) or an IO stream (binary or text).</span>
<span class="sd">    :param format_: Format of the file. If None, the format will be guessed from the</span>
<span class="sd">                    filename extension or the stream content.</span>
<span class="sd">    :return: A SeqFile instance</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">,</span> <span class="n">IO</span><span class="p">],</span> <span class="n">format_</span><span class="p">:</span> <span class="n">_SUPPORTED_FORMATS</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">temp_prefix</span><span class="o">=</span><span class="s1">&#39;seqfile_&#39;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;unknown&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Path</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">format</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="n">format_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">IO</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_from_stream</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_open_func</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">Callable</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">file</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;stdin&#39;</span><span class="p">}:</span>  <span class="c1"># Handle stdin symbol, the rest of the logic should deal with the stream</span>
            <span class="n">file</span> <span class="o">=</span> <span class="n">stdin</span>

        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;read&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">)):</span>  <span class="c1"># --- Handle IO Stream Input ---</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="p">(</span><span class="n">BufferedIOBase</span><span class="p">,</span> <span class="n">RawIOBase</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">buffer</span> <span class="o">:=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;buffer&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span>
                    <span class="n">file</span> <span class="o">=</span> <span class="n">buffer</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">SeqFileError</span><span class="p">(</span><span class="s2">&quot;Input stream is not binary and lacks an accessible binary buffer (.buffer).&quot;</span><span class="p">)</span>

            <span class="c1"># The existing, clever solution: dump the stream to a temp file to make it seekable.</span>
            <span class="k">with</span> <span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="n">temp_prefix</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s1">&#39;.tmp&#39;</span><span class="p">,</span> <span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">temp_f_handle</span><span class="p">:</span>
                <span class="n">copyfileobj</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">temp_f_handle</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">temp_f_handle</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">stem</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_from_stream</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="c1"># Now, we can safely open and guess from the temp file.</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">format</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>  <span class="c1"># xopen handles decompression automatically based on magic numbers</span>
                    <span class="k">with</span> <span class="n">xopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">format</span> <span class="o">=</span> <span class="n">_guess_format_from_handle</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># Clean up the temp file on failure</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="n">e</span>

            <span class="c1"># The suffix for the temp file doesn&#39;t matter since we use xopen&#39;s &#39;magic&#39; method</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_open_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">xopen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;magic&#39;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rt&#39;</span><span class="p">)</span>

        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">)):</span>  <span class="c1"># --- Handle File Path Input ---</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">path</span> <span class="o">=</span> <span class="n">file</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">Path</span><span class="p">)</span> <span class="k">else</span> <span class="n">Path</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">m</span> <span class="o">:=</span> <span class="n">_SEQUENCE_FILE_REGEX</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">name</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">())</span>  <span class="c1"># If format is not provided, guess from extension.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">format</span> <span class="o">=</span> <span class="n">format_</span> <span class="ow">or</span> <span class="nb">next</span><span class="p">(</span><span class="n">fmt</span> <span class="k">for</span> <span class="n">fmt</span> <span class="ow">in</span> <span class="n">get_args</span><span class="p">(</span><span class="n">_SUPPORTED_FORMATS</span><span class="p">)</span> <span class="k">if</span> <span class="n">m</span><span class="p">[</span><span class="n">fmt</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_open_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">xopen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">m</span><span class="p">[</span><span class="s1">&#39;compression&#39;</span><span class="p">]</span> <span class="ow">or</span> <span class="s1">&#39;uncompressed&#39;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rt&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># If no valid extension, try guessing from content</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">format</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="k">with</span> <span class="n">xopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">format</span> <span class="o">=</span> <span class="n">_guess_format_from_handle</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_open_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">xopen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;magic&#39;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rt&#39;</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                         <span class="k">raise</span> <span class="n">SeqFileError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Unsupported file extension and could not guess format for: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span> <span class="kn">from</span><span class="w"> </span><span class="nn">e</span>
                <span class="k">else</span><span class="p">:</span> <span class="c1"># Format was provided, but extension is weird. Trust the user.</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_open_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">xopen</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;magic&#39;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rt&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Input must be a file path (str or Path) or an IO stream, not </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">file</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;SeqFile(</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1">, format=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">format</span><span class="si">}</span><span class="s1">)&#39;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__enter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_handle_open</span><span class="p">()</span>  <span class="c1"># Open the handle when entering context, using the appropriate _open_func</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__exit__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exc_type</span><span class="p">,</span> <span class="n">exc_val</span><span class="p">,</span> <span class="n">exc_tb</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># Ensure the primary handle (_handle) is closed</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_stream</span><span class="p">:</span>  <span class="c1"># Clean up the temporary file *only* if it was created from a stream</span>
            <span class="k">try</span><span class="p">:</span>  <span class="c1"># Use missing_ok=True for robustness against race conditions etc.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">OSError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># Log or warn about failure to delete temp file</span>
                 <span class="n">warn</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Could not delete temporary file </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">SeqFileWarning</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__del__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>  <span class="c1"># Minimal cleanup: try to close the handle and delete temp file if applicable</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># Subject to usual __del__ caveats (timing, interpreter state)</span>
        <span class="k">if</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;_from_stream&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;path&#39;</span><span class="p">):</span>
             <span class="k">try</span><span class="p">:</span>  <span class="c1"># Avoid print/warn in __del__ if possible, or use sys.stderr</span>
                 <span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
             <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                 <span class="k">pass</span> <span class="c1"># Ignore errors in __del__</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_ensure_handle_open</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Internal helper to open/reopen the file via _open_func.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">closed</span><span class="p">:</span>
             <span class="bp">self</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># Close any potentially lingering closed handle first</span>
             <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_open_func</span><span class="p">()</span>  <span class="c1"># Open fresh using the configured function (xopen for paths/streams)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Closes the associated file handle (_handle), if open.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">closed</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                 <span class="k">pass</span>   <span class="c1"># Ignore potential errors during close, especially in __del__ context</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Mark as closed</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">read</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Reads the entire content of the file (decompressed).&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_handle_open</span><span class="p">()</span>  <span class="c1"># Ensure we have a fresh handle from the start</span>
        <span class="n">content</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>  <span class="c1"># Read from the current handle (which was just opened/rewound)</span>
        <span class="c1"># Close handle after full read? Optional, depends on expected usage.</span>
        <span class="c1"># self.close()</span>
        <span class="k">return</span> <span class="n">content</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">rewind</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Resets the file to be read from the beginning.&quot;&quot;&quot;</span>
        <span class="c1"># The simplest way to rewind when using xopen dynamically is to close</span>
        <span class="c1"># the current handle and let _ensure_handle_open get a new one.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># Next call to read() or __iter__() will automatically reopen via _ensure_handle_open()</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="s1">&#39;Record&#39;</span><span class="p">,</span> <span class="s1">&#39;Edge&#39;</span><span class="p">,</span> <span class="s1">&#39;Feature&#39;</span><span class="p">],</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns an iterator (generator) over records in the file.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_handle_open</span><span class="p">()</span> <span class="c1"># Ensure handle is open/reopened</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">format</span><span class="p">:</span>
             <span class="k">raise</span> <span class="n">SeqFileError</span><span class="p">(</span><span class="s2">&quot;Cannot parse file: format is unknown.&quot;</span><span class="p">)</span>
        <span class="k">yield from</span> <span class="n">parse</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">format</span><span class="p">)</span>
        <span class="c1"># Note: Iteration consumes the handle. Re-iteration requires rewind()/re-opening.</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">peek</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the first line and resets to the beginning.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_handle_open</span><span class="p">()</span>
        <span class="n">first_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rewind</span><span class="p">()</span>  <span class="c1"># Rewind by closing and letting the next operation reopen</span>
        <span class="k">return</span> <span class="n">first_line</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="eris.io.SeqFile.__iter__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__iter__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Returns an iterator (generator) over records in the file.</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="s1">&#39;Record&#39;</span><span class="p">,</span> <span class="s1">&#39;Edge&#39;</span><span class="p">,</span> <span class="s1">&#39;Feature&#39;</span><span class="p">],</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns an iterator (generator) over records in the file.&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_handle_open</span><span class="p">()</span> <span class="c1"># Ensure handle is open/reopened</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">format</span><span class="p">:</span>
         <span class="k">raise</span> <span class="n">SeqFileError</span><span class="p">(</span><span class="s2">&quot;Cannot parse file: format is unknown.&quot;</span><span class="p">)</span>
    <span class="k">yield from</span> <span class="n">parse</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">format</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.SeqFile.close" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">close</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Closes the associated file handle (_handle), if open.</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Closes the associated file handle (_handle), if open.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">closed</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
             <span class="k">pass</span>   <span class="c1"># Ignore potential errors during close, especially in __del__ context</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Mark as closed</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.SeqFile.peek" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">peek</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Returns the first line and resets to the beginning.</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">peek</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the first line and resets to the beginning.&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_handle_open</span><span class="p">()</span>
    <span class="n">first_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">rewind</span><span class="p">()</span>  <span class="c1"># Rewind by closing and letting the next operation reopen</span>
    <span class="k">return</span> <span class="n">first_line</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.SeqFile.read" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Reads the entire content of the file (decompressed).</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">read</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Reads the entire content of the file (decompressed).&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_handle_open</span><span class="p">()</span>  <span class="c1"># Ensure we have a fresh handle from the start</span>
    <span class="n">content</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>  <span class="c1"># Read from the current handle (which was just opened/rewound)</span>
    <span class="c1"># Close handle after full read? Optional, depends on expected usage.</span>
    <span class="c1"># self.close()</span>
    <span class="k">return</span> <span class="n">content</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="eris.io.SeqFile.rewind" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">rewind</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Resets the file to be read from the beginning.</p>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">rewind</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Resets the file to be read from the beginning.&quot;&quot;&quot;</span>
    <span class="c1"># The simplest way to rewind when using xopen dynamically is to close</span>
    <span class="c1"># the current handle and let _ensure_handle_open get a new one.</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># Next call to read() or __iter__() will automatically reopen via _ensure_handle_open()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h2 id="eris.io.group_reads" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">group_reads</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Helper function, mostly for CLI, for grouping together read files for the same sample.
Useful for grouping paired-end or hybrid (long + short) readsets.</p>


<details class="notes" open>
  <summary>Notes</summary>
  <ul>
<li>Whilst this function can accept input streams, everything is coerced into ReadFile instances and will be
  written to disk, so this should be taken into consideration before processing many streams.</li>
</ul>
</details>

<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>reads</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-external" title="typing.Iterable" href="https://docs.python.org/3/library/typing.html#typing.Iterable">Iterable</a>[<a class="autorefs autorefs-external" title="typing.Union" href="https://docs.python.org/3/library/typing.html#typing.Union">Union</a>[<a class="autorefs autorefs-external" title="pathlib.Path" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path">Path</a>, <a class="autorefs autorefs-external" href="https://docs.python.org/3/library/stdtypes.html#str">str</a>, <a class="autorefs autorefs-external" title="typing.IO" href="https://docs.python.org/3/library/typing.html#typing.IO">IO</a>, <a class="autorefs autorefs-internal" title="ReadFile (eris.io.ReadFile)" href="#eris.io.ReadFile">ReadFile</a>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An iterable of files as strings or Path objects, ReadFile or IO streams.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Yields:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="ReadSet (eris.io.ReadSet)" href="#eris.io.ReadSet">ReadSet</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ReadSet instances</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">readset</span> <span class="ow">in</span> <span class="n">group_reads</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s1">&#39;reads&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">iterdir</span><span class="p">()):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">readset</span><span class="si">:</span><span class="s1">fasta</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">group_reads</span><span class="p">(</span><span class="n">reads</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="n">Path</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="n">IO</span><span class="p">,</span> <span class="n">ReadFile</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">ReadSet</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function, mostly for CLI, for grouping together read files for the same sample.</span>
<span class="sd">    Useful for grouping paired-end or hybrid (long + short) readsets.</span>

<span class="sd">    Notes:</span>
<span class="sd">        - Whilst this function can accept input streams, everything is coerced into ReadFile instances and will be</span>
<span class="sd">          written to disk, so this should be taken into consideration before processing many streams.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        reads: An iterable of files as strings or Path objects, ReadFile or IO streams.</span>

<span class="sd">    Yields:</span>
<span class="sd">        ReadSet instances</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; from pathlib import Path</span>
<span class="sd">        &gt;&gt;&gt; for readset in group_reads(Path(&#39;reads&#39;).iterdir()):</span>
<span class="sd">        &gt;&gt;&gt;     print(f&#39;{readset:fasta}&#39;, end=&#39;&#39;)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">sample_name</span><span class="p">,</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">grouper</span><span class="p">(((</span><span class="n">ReadFile</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">ReadFile</span><span class="p">)</span> <span class="k">else</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">),</span> <span class="s1">&#39;sample_name&#39;</span><span class="p">):</span>
        <span class="k">yield</span> <span class="n">ReadSet</span><span class="p">(</span><span class="o">*</span><span class="n">files</span><span class="p">,</span> <span class="n">id_</span><span class="o">=</span><span class="n">sample_name</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="eris.io.parse" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="n">format_</span><span class="o">=</span><span class="s1">&#39;guess&#39;</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Simple parser for fasta, gfa and genbank formats,
similar to Biopython <a href="https://biopython.org/docs/latest/api/Bio.SeqIO.html#Bio.SeqIO.parse">https://biopython.org/docs/latest/api/Bio.SeqIO.html#Bio.SeqIO.parse</a></p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>handle</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-external" title="typing.TextIO" href="https://docs.python.org/3/library/typing.html#typing.TextIO">TextIO</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A file handle opened in text-mode / text stream</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>format_</code>
            </td>
            <td>
                  <code><span title="eris.io._SUPPORTED_FORMATS">_SUPPORTED_FORMATS</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The format of the file; must be one of the supported formats or guess by default</p>
              </div>
            </td>
            <td>
                  <code>&#39;guess&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Yields:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-external" title="typing.Union" href="https://docs.python.org/3/library/typing.html#typing.Union">Union</a>[<a class="autorefs autorefs-internal" title="Record (eris.seq.Record)" href="../seq/#eris.seq.Record">Record</a>, <a class="autorefs autorefs-internal" title="Edge (eris.graph.Edge)" href="../graph/#eris.graph.Edge">Edge</a>, <a class="autorefs autorefs-internal" title="Feature (eris.seq.Feature)" href="../seq/#eris.seq.Feature">Feature</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Record, Feature or Edge objects</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>eris/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">:</span> <span class="n">TextIO</span><span class="p">,</span> <span class="n">format_</span><span class="p">:</span> <span class="n">_SUPPORTED_FORMATS</span> <span class="o">=</span> <span class="s1">&#39;guess&#39;</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Generator</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="n">Record</span><span class="p">,</span> <span class="n">Edge</span><span class="p">,</span> <span class="n">Feature</span><span class="p">],</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Simple parser for fasta, gfa and genbank formats,</span>
<span class="sd">    similar to Biopython &lt;https://biopython.org/docs/latest/api/Bio.SeqIO.html#Bio.SeqIO.parse&gt;</span>

<span class="sd">    Parameters:</span>
<span class="sd">        handle: A file handle opened in text-mode / text stream</span>
<span class="sd">        format_: The format of the file; must be one of the supported formats or guess by default</span>

<span class="sd">    Yields:</span>
<span class="sd">        Record, Feature or Edge objects</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">format_</span> <span class="o">==</span> <span class="s1">&#39;guess&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">m</span> <span class="o">:=</span> <span class="n">_SEQUENCE_FILE_REGEX</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">handle</span><span class="o">.</span><span class="n">name</span><span class="p">):</span>
            <span class="n">format_</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">fmt</span> <span class="k">for</span> <span class="n">fmt</span> <span class="ow">in</span> <span class="n">get_args</span><span class="p">(</span><span class="n">_SUPPORTED_FORMATS</span><span class="p">)</span> <span class="k">if</span> <span class="n">m</span><span class="p">[</span><span class="n">fmt</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">SeqFileError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Unsupported SeqFile format or extension: </span><span class="si">{</span><span class="n">handle</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">parser</span> <span class="o">:=</span> <span class="p">{</span><span class="s1">&#39;fasta&#39;</span><span class="p">:</span> <span class="n">_parse_fasta</span><span class="p">,</span> <span class="s1">&#39;gfa&#39;</span><span class="p">:</span> <span class="n">_parse_gfa</span><span class="p">,</span> <span class="s1">&#39;genbank&#39;</span><span class="p">:</span> <span class="n">_parse_genbank</span><span class="p">,</span> <span class="s1">&#39;fastq&#39;</span><span class="p">:</span> <span class="n">_parse_fastq</span><span class="p">,</span>
                  <span class="s1">&#39;gff&#39;</span><span class="p">:</span> <span class="n">_parse_gff</span><span class="p">,</span> <span class="s1">&#39;bed&#39;</span><span class="p">:</span> <span class="n">_parse_bed</span><span class="p">}</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">format_</span><span class="p">):</span>
        <span class="k">yield from</span> <span class="n">parser</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Format &quot;</span><span class="si">{</span><span class="n">format_</span><span class="si">}</span><span class="s1">&quot; not supported&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>







  
  






                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["toc.follow", "content.action.edit"], "search": "../../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.50899def.min.js"></script>
      
    
  </body>
</html>